third-party/realdds/src/dds-device-server.cpp

```diff
@@ -1,186 +1,137 @@
-}
-
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2022 Intel Corporation. All Rights Reserved.
+
+#include <realdds/dds-device-server.h>
+
+#include <realdds/dds-participant.h>
+#include <realdds/dds-publisher.h>
+#include <realdds/dds-stream-server.h>
+#include <realdds/dds-stream-profile.h>
+#include <realdds/dds-notification-server.h>
+#include <realdds/dds-utilities.h>
+#include <realdds/topics/device-info/device-info-msg.h>
+#include <realdds/topics/image/image-msg.h>
+#include <realdds/topics/notification/notification-msg.h>
+#include <realdds/dds-topic.h>
+#include <realdds/dds-topic-writer.h>
+
+#include <fastdds/dds/topic/Topic.hpp>
+
+#include <third-party/json.hpp>
+using nlohmann::json;
+
+
+using namespace eprosima::fastdds::dds;
+using namespace realdds;
+
+
+dds_device_server::dds_device_server( std::shared_ptr< dds_participant > const & participant,
+                                      const std::string & topic_root )
+    : _publisher( std::make_shared< dds_publisher >( participant ))
+    , _topic_root( topic_root )
+{
+    LOG_DEBUG( "device server created @ '" << _topic_root << "'" );
+}
+
+
+dds_device_server::~dds_device_server()
+{
+    _stream_name_to_server.clear();
+    LOG_DEBUG( "device server deleted @ '" << _topic_root << "'" );
+}
+
+
+void dds_device_server::start_streaming( const std::string & stream_name,
+                                         const image_header & header )
+{
+    auto it = _stream_name_to_server.find( stream_name );
+    if( it == _stream_name_to_server.end() )
+        DDS_THROW( runtime_error, "stream '" + stream_name + "' does not exist" );
+    auto& stream = it->second;
+    stream->start_streaming( header );
+}
+
+
+void dds_device_server::publish_image( const std::string & stream_name, const uint8_t * data, size_t size )
+{
+    auto it = _stream_name_to_server.find( stream_name );
+    if( it == _stream_name_to_server.end() )
+        DDS_THROW( runtime_error, "stream '" + stream_name + "' does not exist" );
+    auto & stream = it->second;
+    if( ! stream->is_streaming() )
+        DDS_THROW( runtime_error, "stream '" + stream_name + "' is not streaming" );
+
+    stream->publish_image( data, size );
+}
+
+
+static void on_discovery_device_header( size_t const n_streams, dds_notification_server & notifications )
+{
+    topics::notification notification( json{
+        { "id", "device-header" },
+        { "n-streams", n_streams },
+    } );
+    notifications.add_discovery_notification( std::move( notification ) );
+}
+
+
+static void on_discovery_stream_header( std::shared_ptr< dds_stream_server > const & stream,
+                                        dds_notification_server & notifications )
+{
+    nlohmann::ordered_json profiles = {};
+    for( auto & sp : stream->profiles() )
+        profiles += sp->to_json();
+    json msg = {
+        { "id", "stream-header" },
+        { "type", stream->profiles().front()->type_to_string() },
```

Comment:
To support rs2 features we will need to add a `dds_stream_xxx` type system and a conversion back to `rs2_stream`+index (like we convert `dds_stream_format` to and from `rs2_format`)

This will be able to replace the changes I made in my PR ([#11045], (https://github.com/IntelRealSense/librealsense/pull/11045), you did add some comments about it)

Reply:
Will also need some factory to create all the different stream types

Reply:
Once both our PRs are in, I'll add all the other classes